lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details
+
+r"""Interface library for the Prometheus Pushgateway.
+
+This library wraps a relation endpoint using the `pushgwateway` interface
+and exposes an API for forwarding metrics to Prometheus.
+
+
+## Getting Started
+
+To get started using the library, you just need to fetch the library using `charmcraft`.
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.prometheus_pushgateway_k8s.v0.pushgateway
+```
+
+In the `metadata.yaml` of the charm, add the following:
+
+```yaml
+requires:
+  your-relation-name:
+    interface: pushgateway
+```
+
+In the source of your charm, first import the interface:
+
+```
+from charms.prometheus_pushgateway_k8s.v0.pushgateway import PrometheusPushgatewayRequirer
+````
+
+Instantiate the object in your charm's `__init__`, like so:
+
+```
+from charms.prometheus_pushgateway_k8s.v0.pushgateway import PrometheusPushgatewayRequirer
+from ops.charm import CharmBase
+
+class MyCharm(CharmBase):
+    def __init__(...):
+         ...
+        self.pushgateway_requirer = PrometheusPushgatewayRequirer(self, "your-relation-name")
```

Comment:
Is this intended to be used as part of `update-status`?
Otherwise the charm may never get to call `send-metric` after startup.

Reply:
No intention to be used on any specific event. The user may want to send a metric on how much it take to do something when config changed, for example.

I don't understand why you say the charm may never get to call `send-metric`...

Reply:
Pushgateway seems like a functionality that a workload would need.
The only time a charm could push anything is when a hook fires, and the only periodic hook we have is update-status.
But future apps we charm could certainly make use of `requirer._pushgateway_url`, in which case: shouldn't it be public?

Reply:
For sure this will be used in other situations than `update-status`. Think for example a charm action to make a backup of the service's data. It totally could measure how long it takes and report it at the end of the action.

Regarding the usage of the pushgateway's URL, it may happen in the future. In that case we could make the attribute public. OTOH, once is public, we could never make it private again. So I'd leave it private for now unless you have a compelling case *now*.

Reply:
That's fine. I didn't realize this was primarily intended for in-charm usage.